content/src/content/jcr_root/apps/core/wcm/components/commons/datalayer/clientlibs/js/datalayer.js

```diff
@@ -0,0 +1,264 @@
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+(function() {
+    "use strict";
+
+    /* eslint no-console: "off" */
+
+    var CHANGE_EVENT = "datalayer:change";
+
+    // Initializes the data layer
+    function init() {
+        window.dataLayer = window.dataLayer || [];
+        window.dataLayer.state = {};
+        window.dataLayer._listeners = [];
+        populateDataLayer();
+        handleEventsBeforeScriptLoad(window.dataLayer);
+        overridePush();
+        populateDataLayerAfterOverride();
+        console.log("data layer script initialized");
+    }
+
+    // Augments the push function (only for the data layer object) to also handle the event
+    function overridePush() {
+        window.dataLayer.push = function() {
+            var pushArguments = arguments;
+            var filteredArguments = arguments;
+            Object.keys(pushArguments).forEach(function(key) {
+                var event = pushArguments[key];
+                handleEvent(event);
+                // filter out event listeners
+                if (event.handler) {
+                    delete filteredArguments[key];
+                }
+            });
+            if (filteredArguments[0]) {
+                return Array.prototype.push.apply(this, filteredArguments);
+            }
+        };
+    }
+
+    function handleEventsBeforeScriptLoad(dataLayer) {
+        dataLayer.forEach(function(event, idx) {
+            // remove event listeners that were defined before the script load.
+            if (event.handler) {
+                dataLayer.splice(idx, 1);
+            }
+            handleEvent(event);
+        });
+    }
+
+    function handleEvent(event) {
+        if (!event) {
+            return;
+        }
+        if (event.data) {
+            updateState(window.dataLayer.state, event.data);
+            triggerListeners(event);
+        } else if (event.handler) {
+            registerListener(event);
+            triggerListener(event);
+        }
+    }
+
+    // trigger the listener on all previous events matching the listener
+    function triggerListener(listener) {
+        window.dataLayer.forEach(function(event) {
+            if (listener.on === CHANGE_EVENT || listener.on === event.type) {
+                listener.handler(event);
+            }
+        });
+    }
+
+    function registerListener(listener) {
+        // add the listener to datalayer._listeners
+        window.dataLayer._listeners.push(listener);
+        console.log("register an event listener: " + listener.on);
+    }
+
+    function triggerListeners(event) {
+        // loop over all the listeners
+        // when a match is found, execute the handler
+        Object.keys(window.dataLayer._listeners).forEach(function(key) {
+            var listener = window.dataLayer._listeners[key];
+            if (listener.on === CHANGE_EVENT || listener.on === event.type) {
```

Comment:
wondering what order we should fire these in:
1. `event.type` first, then `CHANGE_EVENT` events
2. `CHANGE_EVENT` events first, then `event.type` events
3. The order they were registered regardless of type